designs/2023-only-run-reporting-rules/README.md

```diff
@@ -0,0 +1,129 @@
+
+From an API perspective, this would be implemented by a predicate function that acts as
+a filter on which rules should be run. The predicate would take the rule configuration,
+and return whether the rule should be run. For simplicity, this function should not be
+given rules marked as `off`, as if this function handled existing behavior then users of the
+API would have to mimic that when attempting to extend it.
+
+The flag would be implemented by passing a predicate function that only returns true for
+rules that have been set to `error`, thus filtering out any that are marked as `warn`.
+
+The check for unused `eslint-disable` directives (`--report-unused-disable-directives`)
+should continue to mark `warn` rules as used, even when running with `--skip-warnings`.
+
+The `--max-warnings` flag should disable the `--skip-warnings` flag, as it requires that
+`warn` rules are run to know the current number of reported warnings.
+
+## Documentation
+
+Both the flag and API would make sense to document. As they are relatively simple additions,
+they would be documented inline with the rest of the CLI and API documentation. Due to the
+potentially major performance improvements for many setups, I feel mentioning it in the
+release blog post, alongside the potential benefits, would be a good idea to spread awareness.
+
+## Drawbacks
+
+This change adds a new command line flag, as well as a new API method. This adds further
+maintenance burden. However, the implementation is relatively simple, and the benefits
+are significant.
+
+Another potential drawback around this specific implementation is that the existence of
+both `--quiet` and `--skip-warnings` might be confusing to some users, however this can be
+partially alleviated by updating the `--quiet` docs to mention that warnings will still be
+run, and to use `--skip-warnings` instead to not run them.
+
+## Backwards Compatibility Analysis
+
+The proposed solution does not cause any compatibility issues, as all functionality is
+implemented via new APIs and command line flags. Existing users will be unaffected
+by this change.
+
+## Alternatives
+
+### Implement this into `--quiet` instead of a new flag
+
+This alternative would still require the same API changes, but would instead implement
+this behavior into the `--quiet` flag. This would be a breaking change, as it would
+change the behavior of `--quiet` to not run `warn` rules. The upside of this alternative
+would be that it's less confusing to users. This would presumably need to wait for ESLint 9.
+
+Another downside of this alternative is that it would introduce confusion around the
+`--max-warnings` flag and other checks that require warnings to run. Currently the quiet
+flag works fine alongside these. If we were to implement this into the quiet flag, we would
+have to either break that functionality, or have the quiet flag only sometimes stop actual
+running depending on what other flags are in use, which is confusing and unclear.
```

Comment:
Since this alternative is a breaking change, I think we could reconsider interactions between flags as needed, and even throw an exception when multiple flags are provided that don't make sense to be specified together.

It would help to create a table explaining the current behavior with each combination of flags and also how the behavior would change for each combination (and whether it would still be allowed).

Reply:
Agree. I also think that as long as we're explicit with error messages, we are okay to make such changes.

Reply:
> and whether it would still be allowed

I think it should, at least in the case of `--quiet` and `--max-warning`, since you logically expect warnings to be skipped (right now are not) but user should be notified that the behavior changed and that running `--max-warnings` flag alongside `--quiet` is redundant.